pennylane/_qubit_device.py

```diff
@@ -29,100 +30,164 @@
+        * :meth:`~.generate_samples`
+
+        * :meth:`~.probability`
 
         Args:
-            queue (Iterable[~.operation.Operation]): operations to execute on the device
-            observables (Iterable[~.operation.Observable]): observables to measure and return
-            parameters (dict[int->list[ParameterDependency]]): Mapping from free parameter index to the list of
-                :class:`Operations <pennylane.operation.Operation>` (in the queue) that depend on it.
+            circuit (~.CircuitGraph): circuit to execute on the device
 
         Raises:
             QuantumFunctionError: if the value of :attr:`~.Observable.return_type` is not supported
 
         Returns:
             array[float]: measured value(s)
         """
-        if parameters is None:
-            parameters = {}
-
-        self.check_validity(queue, observables)
-        self._op_queue = queue
-        self._obs_queue = observables
-        self._parameters = {}
-        self._parameters.update(parameters)
+        self.check_validity(circuit.operations, circuit.observables)
 
         with self.execution_context():
-            self.pre_apply()
-
-            for operation in queue:
-                # Pass instances directly
-                self.apply(operation)
-
-            self.rotate_basis(observables)
-
-            self.post_apply()
+            # determine operations required to diagonalize observables
+            rotations = self.rotate_basis(circuit.observables)
 
-            self.pre_measure()
+            # apply all circuit operations
+            self.apply(circuit.operations, rotations)
 
-            results = self.statistics(observables)
+            # generate computational basis samples
+            self.generate_samples()
 
-            self.post_measure()
-
-            self._op_queue = None
-            self._obs_queue = None
-            self._parameters = None
+            # compute the required statistics
+            results = self.statistics(circuit.observables)
 
             # Ensures that a combination with sample does not put
             # expvals and vars in superfluous arrays
-            sample_return_types = (obs.return_type is Sample for obs in observables)
+            sample_return_types = (obs.return_type is Sample for obs in circuit.observables)
             if any(sample_return_types) and not all(sample_return_types):
                 return self._asarray(results, dtype="object")
 
             return self._asarray(results)
 
+    @abc.abstractmethod
+    def apply(self, operations, rotations=None, **kwargs):
+        """Apply quantum operations, and execute the quantum circuit.
+
+        Args:
+            operations (list[~.Operation]): operations to apply to the device
+            rotations (list[~.Operation]): operations that rotate the circuit
+                pre-measurement into the eigenbasis of the observables.
+        """
+
+    def rotate_basis(self, observables):
```

Comment:
I would consider renaming this to something like `diagonalizing_rotations` or so. Right now, the imperative "rotate basis" suggests that the rotation takes place inside this function whereas it only returns the gates necessary for diagonalization.

Reply:
Could make it `diagonalizing_gates` to match the equivalent `Operation` method?

Reply:
I agree that ``rotate_basis`` would imply currently, that there are rotations taking place. ``diagonalizing_gates`` could make sense given that we are basically getting all the ``diagonalizing_gates`` of the observables and creating a ``list`` out of it.

Reply:
We have moved `rotate_basis` to a property `CircuitGraph.diagonalizing_gates`, which makes much more sense :)